lib/netext/httpext/request.go

```diff
@@ -0,0 +1,373 @@
+			preq.Req.Header.Set(digest.KEY_AUTHORIZATION, authorization)
+		}
+	}
+
+	debugRequest(state, preq.Req, "Request")
+	res, resErr := client.Do(preq.Req.WithContext(ctx))
+	debugResponse(state, res, "Response")
+	resp.Error = tracerTransport.errorMsg
+	resp.ErrorCode = int(tracerTransport.errorCode)
+	if resErr == nil && res != nil {
+		switch res.Header.Get("Content-Encoding") {
+		case "deflate":
+			res.Body, resErr = zlib.NewReader(res.Body)
+		case "gzip":
+			res.Body, resErr = gzip.NewReader(res.Body)
+		}
+	}
+	if resErr == nil && res != nil {
+		if preq.ResponseType == ResponseTypeNone {
+			_, err := io.Copy(ioutil.Discard, res.Body)
+			if err != nil && err != io.EOF {
+				resErr = err
+			}
+			resp.Body = nil
+		} else {
+			// Binary or string
+			buf := state.BPool.Get()
+			buf.Reset()
+			defer state.BPool.Put(buf)
+			_, err := io.Copy(buf, res.Body)
+			if err != nil && err != io.EOF {
+				resErr = err
+			}
+
+			switch preq.ResponseType {
+			case ResponseTypeText:
+				resp.Body = buf.String()
+			case ResponseTypeBinary:
+				resp.Body = buf.Bytes()
+			default:
+				resErr = fmt.Errorf("unknown responseType %s", preq.ResponseType)
+			}
+		}
+		_ = res.Body.Close()
+	}
+
+	trail := tracerTransport.GetTrail()
+
+	if trail.ConnRemoteAddr != nil {
+		remoteHost, remotePortStr, _ := net.SplitHostPort(trail.ConnRemoteAddr.String())
+		remotePort, _ := strconv.Atoi(remotePortStr)
+		resp.RemoteIP = remoteHost
+		resp.RemotePort = remotePort
+	}
+	resp.Timings = ResponseTimings{
+		Duration:       stats.D(trail.Duration),
+		Blocked:        stats.D(trail.Blocked),
+		Connecting:     stats.D(trail.Connecting),
+		TLSHandshaking: stats.D(trail.TLSHandshaking),
+		Sending:        stats.D(trail.Sending),
+		Waiting:        stats.D(trail.Waiting),
+		Receiving:      stats.D(trail.Receiving),
+	}
+
+	if resErr == nil {
+		if preq.ActiveJar != nil {
+			if rc := res.Cookies(); len(rc) > 0 {
+				preq.ActiveJar.SetCookies(res.Request.URL, rc)
+			}
+		}
+
+		resp.URL = res.Request.URL.String()
+		resp.Status = res.StatusCode
+		resp.Proto = res.Proto
+
+		if res.TLS != nil {
+			resp.setTLSInfo(res.TLS)
+		}
+
+		resp.Headers = make(map[string]string, len(res.Header))
+		for k, vs := range res.Header {
+			resp.Headers[k] = strings.Join(vs, ", ")
+		}
+
+		resCookies := res.Cookies()
+		resp.Cookies = make(map[string][]*HTTPCookie, len(resCookies))
+		for _, c := range resCookies {
+			resp.Cookies[c.Name] = append(resp.Cookies[c.Name], &HTTPCookie{
+				Name:     c.Name,
+				Value:    c.Value,
+				Domain:   c.Domain,
+				Path:     c.Path,
+				HTTPOnly: c.HttpOnly,
+				Secure:   c.Secure,
+				MaxAge:   c.MaxAge,
+				Expires:  c.Expires.UnixNano() / 1000000,
+			})
+		}
+	}
+
+	if resErr != nil {
+		// Do *not* log errors about the contex being cancelled.
+		select {
+		case <-ctx.Done():
+		default:
+			state.Logger.WithField("error", resErr).Warn("Request Failed")
+		}
+
+		if preq.Throw || resp.Error == "" {
```

Comment:
I am confused, why did you add `|| resp.Error == ""` to this check?

Reply:
As per the offline discussion we had: if we get an error that is not from the RoundTripper we should throw an exception in goja this is basically the same behaviour as if `throw` is set to true. I check `resp.Error` as I should have previously set it to the error from the RoundTripper if it is empty but we have an error it means the error is not from the RoundTripper and we want to throw an exception in goja.

Reply:
Can you add something like this (maybe with a few commas :wink: ) as a comment inside of the code, since it totally isn't obvious from just reading the code.